### PR TITLE
Refactor train.py

### PR DESCRIPTION
**Pull request by isaaccorley**
This PR refactors `train.py` to use `hydra.utils.instantiate` to define the trainer and datamodule inside the config file without having to do any config file magic. This allows a user to define a config like below:

```yaml
trainer:
  _target_: pytorch_lightning.Trainer
  accelerator: gpu
  devices: 1
  min_epochs: 15
  max_epochs: 40

module:
  _target_: torchgeo.trainers.ClassificationTask
  loss: "ce"
  model: "resnet18"
  learning_rate: 1e-3
  learning_rate_schedule_patience: 6
  weights: null
  in_channels: 3
  num_classes: 45

datamodule:
  _target_: torchgeo.datamodules.RESISC45DataModule
  root: "data/resisc45"
  batch_size: 128
  num_workers: 4
```

### CODE REVIEW

**Comment by adamjstewart** [comment]
We talked about this in the meeting today, but we're probably going to try to combine pretrain/train/evaluate/predict into a single main.py that can handle all of the above, and store things like datamodule and trainer classes in the YAML config file.

**Comment by adamjstewart** [comment]
Actually, we could name it `torchgeo/__main__.py` so that users can run `python -m torchgeo --train --datamodule=... --task=...`. This might be sufficient to close #228.

File: train.py [hunk 1]
```diff
@@ -44,32 +43,39 @@
...
-    "naipchesapeake": (SemanticSegmentationTask, NAIPChesapeakeDataModule),
-    "nasa_marine_debris": (ObjectDetectionTask, NASAMarineDebrisDataModule),
-    "potsdam2d": (SemanticSegmentationTask, Potsdam2DDataModule),
-    "resisc45": (ClassificationTask, RESISC45DataModule),
-    "sen12ms": (SemanticSegmentationTask, SEN12MSDataModule),
-    "so2sat": (ClassificationTask, So2SatDataModule),
-    "spacenet1": (SemanticSegmentationTask, SpaceNet1DataModule),
-    "ucmerced": (ClassificationTask, UCMercedDataModule),
-    "vaihingen2d": (SemanticSegmentationTask, Vaihingen2DDataModule),
+DATASET_TO_DATAMODULES_MAPPING: Dict[str, Type[LightningDataModule]] = {
```
Comment (adamjstewart) [inline train.py:49]: Can we automatically infer these? They follow a consistent naming scheme.

Comment (isaaccorley) [reply train.py:49]: How would you infer these? I mentioned using hydra to instantiate from a config. Do you have any other suggestions?

Comment (adamjstewart) [reply train.py:49]: Can we use the full CamelCase name in the config? I'm fine with using hydra.

Comment (isaaccorley) [reply train.py]: It's possible to use either `torchgeo.trainers.ClassificationTask` or `ClassificationTask` in the config

Comment (isaaccorley) [reply train.py]: The next question is should we make a user only have to define `ClassificationTask` or prefer them to explicitly define the full path like `torchgeo.trainers.ClassificationTask`

Comment (adamjstewart) [reply train.py]: Ideally either one would work. The former because it's short, and the latter because a user may want to define their own trainer/datamodule. But if it's too much code, I would prefer the former for now.

Comment (isaaccorley) [reply train.py]: I'm going to use the latter because it basically will remove code but also not require us to always update the `train.py` script when we add a new datamodule, also allowing a user to define their own trainer/datamodule.

Comment (adamjstewart) [reply train.py]: Why would it require us to update `train.py`? Couldn't we just use `getattr(torchgeo.datamodules, "LandCoverAIDataModule")`?

**Comment by calebrob6** [comment]
I guess we can get rid of evaluate now that we have testing in `train.py`

File: train.py (hunk 1 above)
Comment (isaaccorley) [reply train.py]: Yes that's right.

**Comment by adamjstewart** [comment]
> ```yaml
> trainer:
>   _target_: pytorch_lightning.Trainer
> ```

It should actually be `lightning.pytorch.Trainer`, we no longer uses pytorch-lightning.

Why don't any of the config files use Trainer?

**Comment by adamjstewart** [comment]
Still need to look into hydra, but I'm wondering if we can remove our omegaconf dep and replace it with hydra? Trying to minimize the deps we have to keep track of.

**Comment by isaaccorley** [comment]
I was having them inherit from conf/defaults.yaml but you're right it's a bit confusing so it's probably better explicitly define it in each config

File: conf/bigearthnet.yaml [hunk 2]
```diff
@@ -1,22 +1,24 @@
...
+datamodule:
+  _target_: torchgeo.datamodules.BigEarthNetDataModule
+  root: "data/bigearthnet"
+  bands: "all"
+  num_classes: ${experiment.module.num_classes}
+  batch_size: 128
+  num_workers: 4
+
 trainer:
+  _target_: lightning.Trainer
```
Comment (adamjstewart) [inline conf/bigearthnet.yaml:20]:
```suggestion
  _target_: lightning.pytorch.Trainer
```
Some classes are aliased and some aren't. I like to be consistent and explicit.

File: requirements/min-reqs.old [hunk 3]
```diff
@@ -46,6 +46,7 @@ pydocstyle[toml]==6.1.0
 pyupgrade==2.8.0
 
 # tests
+hydra-core==1.1.0
 mypy==0.900
 nbmake==1.3.3
 omegaconf==2.1.0
```
Comment (adamjstewart) [inline requirements/min-reqs.old:52]: Have you found a way to remove the omegaconf dep?

File: setup.cfg [hunk 4]
```diff
@@ -117,6 +117,8 @@ style =
     # pyupgrade 2.8+ required for --py39-plus flag
     pyupgrade>=2.8,<4
 tests =
+    # hydra-core 1.1.0+ required for recursive instantiation
```
Comment (adamjstewart) [inline setup.cfg:120]: Are we using recursive instantiation anywhere?

File: train.py [hunk 5]
```diff
@@ -6,70 +6,17 @@
...
-    "nasa_marine_debris": (ObjectDetectionTask, NASAMarineDebrisDataModule),
-    "potsdam2d": (SemanticSegmentationTask, Potsdam2DDataModule),
-    "resisc45": (ClassificationTask, RESISC45DataModule),
-    "sen12ms": (SemanticSegmentationTask, SEN12MSDataModule),
-    "so2sat": (ClassificationTask, So2SatDataModule),
-    "spacenet1": (SemanticSegmentationTask, SpaceNet1DataModule),
-    "ucmerced": (ClassificationTask, UCMercedDataModule),
-    "vaihingen2d": (SemanticSegmentationTask, Vaihingen2DDataModule),
-}
+from torchgeo import datamodules, trainers  # noqa: F401
```
Comment (adamjstewart) [inline train.py:18]: Why is this needed?

File: train.py [hunk 6]
```diff
@@ -202,18 +114,20 @@ def main(conf: DictConfig) -> None:
...
-    ######################################
-    # Run experiment
-    ######################################
+    # Train
     trainer.fit(model=task, datamodule=datamodule)
-    trainer.test(ckpt_path="best", datamodule=datamodule)
+
+    # Test
+    if hasattr(task, "test_step") and hasattr(datamodule, "test_dataloader"):
+        trainer.test(ckpt_path="best", datamodule=datamodule)
```
Comment (adamjstewart) [inline train.py:130]: Unfortunately, all trainers have a test_step (might be a no-op) and all datamodules have a test_dataloader (might raise an error). I would instead use:
```suggestion
    try:
        trainer.test(ckpt_path="best", datamodule=datamodule)
    except MisconfigurationException:
        pass
```
Can add more exceptions if other errors might be raised

File: setup.cfg (hunk 4 above)
Comment (isaaccorley) [reply setup.cfg:120]: Not currently but a user is going to want this. I'll add it to the test config. Recursive instantiation allows you to define the callbacks, logger, augmentations inside the config file itself.

File: train.py (hunk 6 above)
Comment (isaaccorley) [reply train.py:130]: We could also check if test_dataset=None as well but yeah I think your solution works best.

File: train.py (hunk 5 above)
Comment (isaaccorley) [reply train.py:18]: Since importing torchgeo doesn't import the datamodules, trainers, etc. Hydra will try to instantiate the module and datamodules objects but they won't be defined. So we need to import datamodules and trainers. Simple tradeoff imo.

Comment (isaaccorley) [reply train.py:18]: Actually this is solved because I have to import MisconfigurationException from torchgeo.datamodules

File: requirements/min-reqs.old (hunk 3 above)
Comment (isaaccorley) [reply requirements/min-reqs.old]: Just removed. It can be imported like `from hydra.compose import OmegaConf, DictConfig` instead.

Comment (isaaccorley) [reply requirements/min-reqs.old]: Had to upgrade from `hydra-core>=1.1.0` to `hydra-core>=1.1.1` for this

File: tests/conf/trainer.yaml [hunk 7]
```diff
@@ -0,0 +1,26 @@
+trainer:
```
Comment (isaaccorley) [inline tests/conf/trainer.yaml:1]: @adamjstewart Here's an example of recursive instantiation using hydra. This lets you define callbacks, loggers, etc inside the config itself and not having to edit the actual `train.py` script to change these.

File: train.py (hunk 6 above)
Comment (adamjstewart) [reply train.py]: That won't work either because it could be using `self.dataset` instead of `self.test_dataset`.

**Review by adamjstewart — COMMENTED** [review]
Like our existing code, there are a ton of discrepancies as to whether we should call a LightningModule a module, model, task, or trainer. I won't comment on those since we never came to a conclusion on #996.

Config files look amazing! Now just need to decide how closely we can align our training script to LightningCLI.

File: requirements/tests.txt [hunk 8]
```diff
@@ -1,7 +1,7 @@
 # tests
+hydra-core>=1.1.1
```
[the remark this answers is not in the document]
Comment (isaaccorley) [reply requirements/tests.txt:2]: Can you give more context to this?

[the remark this answers is not in the document]
Comment (adamjstewart) [reply requirements/tests.txt]: Files in `requirements` are used for CI, not for installation. We pin all dependencies in these files and manage them with dependabot. You'll notice that all other deps in this file are pinned to the latest version.

File: train.py
[the remark this answers is not in the document]
Comment (isaaccorley) [reply train.py:1]: That seems like a pretty large refactor and I think we should do that in a separate PR. This PR currently refactors train.py so that one can more easily use it with all available or custom modules and datamodules.

File: setup.cfg (hunk 4 above)
Comment (adamjstewart) [inline setup.cfg:120]: What's the minimum version that works without recursive instantiation? Users can install a newer version if they want to, but if older versions work for most cases then I think we should set our minimum to that.

**Comment by adamjstewart** [comment]
I think everything looks good to me. Just want to further minimize the hydra version to the lowest version that gets our tests to pass.

**Review by adamjstewart — APPROVED** [review]

**Comment by calebrob6** [comment]
Pausing a bit to think here. Just so I understand correctly -- the reason we're bringing in hydra as a dependency is for `instantiate`? And this is to 1.) solve the problem of needing to maintain an explicit list of "name" --> (class, task) in `train.py` which 2.) makes `train.py` a lot more flexible?

**Comment by adamjstewart** [comment]
Correct. This also lets users create their own datamodule and/or trainer and use it with `train.py` without having to edit `train.py`. We're planning on moving `train.py` to `torchgeo/__main__.py` so that it gets installed and users can use `python -m torchgeo fit config_file=...`. This will become even more important when we do that.

**Comment by calebrob6** [comment]
Okay cool! I don't see why `train.py` needs to be moved to `torchgeo/__main__.py` actually, but I think that's a discussion for another day.

**Comment by adamjstewart** [comment]
It relates to #228, just provides an easy way for people to `pip install torchgeo` and user our `train.py` script.

**Comment by adamjstewart** [comment]
Yep, I'm planning on integrating this and adding a torchgeo command in the next release.
